### PR TITLE
005 competencies per band

### PR DESCRIPTION
**Pull request by mactrz**
As a Kainos employee
I want to find a list of Kainos Competencies expected of an employee within each band level
So that I know the competencies expected of an employee performing at that band level

Tech notes:
- Competency table
- API endpoint GET /competency/:id

### CODE REVIEW

File: src/main/java/com/kainos/ea/backend/controllers/BandCompetencyController.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+import java.util.List;
+
+@Controller
+@RequestMapping(path="/")
+public class BandCompetencyController {
+
+    private BandCompetencyService bandCompetencyService;
+
+    public BandCompetencyController(BandCompetencyService bandCompetencyService) {
```
Comment (sebastianczajkowski) [inline src/main/java/com/kainos/ea/backend/controllers/BandCompetencyController.java:19]: Needs an @Autowired on line 18

File: src/main/java/com/kainos/ea/backend/controllers/BandController.java [hunk 2]
```diff
@@ -1,23 +1,24 @@
...
 
 @Controller
-@RequestMapping(path="/")
-public class TestController {
+@RequestMapping(path = "/band")
+public class BandController {
 
     @Autowired
-    TestService testService;
+    BandService bandService;
```
Comment (sebastianczajkowski) [inline src/main/java/com/kainos/ea/backend/controllers/BandController.java]: Needs a constructor with @Autowired

File: src/main/java/com/kainos/ea/backend/controllers/BandCompetencyController.java (hunk 1 above)
Comment (mactrz) [reply src/main/java/com/kainos/ea/backend/controllers/BandCompetencyController.java:21]: Added

File: src/main/java/com/kainos/ea/backend/controllers/BandController.java (hunk 2 above)
Comment (mactrz) [reply src/main/java/com/kainos/ea/backend/controllers/BandController.java:20]: Added

File: src/main/java/com/kainos/ea/backend/repositories/BandCompetencyRepository.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package com.kainos.ea.backend.repositories;
+
+import com.kainos.ea.backend.models.BandCompetency;
+import org.springframework.data.repository.CrudRepository;
+
+import java.util.List;
+
+public interface BandCompetencyRepository extends CrudRepository<BandCompetency, String> {
+
+    public List<BandCompetency> findAllByBand_BandName(String bandName);
```
Comment (sebastianczajkowski) [inline src/main/java/com/kainos/ea/backend/repositories/BandCompetencyRepository.java:10]: Is the underscore necessary?

File: src/main/java/com/kainos/ea/backend/services/BandCompetencyService.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+import java.util.List;
+
+
+@Service
+public class BandCompetencyService {
+
+    private BandCompetencyRepository bandCompetencyRepository;
+
+    public BandCompetencyService(BandCompetencyRepository bandCompetencyRepository) {
```
Comment (sebastianczajkowski) [inline src/main/java/com/kainos/ea/backend/services/BandCompetencyService.java:15]: Needs an @Autowired if I'm not mistaken

File: src/test/java/com/kainos/ea/backend/controllers/BandCompetencyControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    @Mock
+    private BandCompetencyService bandCompetencyService;
+
+    @Test
+    public void when_querying_getCompetenciesByBand_expect_ServiceCalledPassback(){
+        List<BandCompetency> bandCompetencies = List.of(new BandCompetency());
+        Mockito.when(bandCompetencyService.getCompetenciesByBand("string")).thenReturn(bandCompetencies);
+        BandCompetencyController bandCompetencyController = new BandCompetencyController(bandCompetencyService);
+
+        Iterable<BandCompetency> results = bandCompetencyController.getCompetenciesByBand("string");
```
Comment (sebastianczajkowski) [inline src/test/java/com/kainos/ea/backend/controllers/BandCompetencyControllerTest.java:30]: It should be a `List` I think

File: src/test/java/com/kainos/ea/backend/services/BandCompetencyServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    @Mock
+    private BandCompetencyRepository bandCompetencyRepository;
+
+    @Test
+    public void when_querying_getCompetenciesByBand_expect_ServiceCalledPassback(){
+        List<BandCompetency> bandCompetencies = List.of(new BandCompetency());
+        Mockito.when(bandCompetencyRepository.findAllByBand_BandName("string")).thenReturn(bandCompetencies);
+        BandCompetencyService bandCompetencyService = new BandCompetencyService(bandCompetencyRepository);
+
+        Iterable<BandCompetency> results = bandCompetencyService.getCompetenciesByBand("string");
```
Comment (sebastianczajkowski) [inline src/test/java/com/kainos/ea/backend/services/BandCompetencyServiceTest.java:27]: Should be a `List` I think

**Review by sebastianczajkowski — CHANGES_REQUESTED** [review]
I left some comments

File: src/main/java/com/kainos/ea/backend/repositories/BandCompetencyRepository.java (hunk 3 above)
Comment (mactrz) [reply src/main/java/com/kainos/ea/backend/repositories/BandCompetencyRepository.java:10]: https://stackoverflow.com/a/55579268
I got it from here

**Review by sebastianczajkowski — APPROVED** [review]
Looks good

File: src/main/java/com/kainos/ea/backend/services/BandCompetencyService.java (hunk 4 above)
Comment (mactrz) [reply src/main/java/com/kainos/ea/backend/services/BandCompetencyService.java:15]: Added

File: src/test/java/com/kainos/ea/backend/services/BandCompetencyServiceTest.java (hunk 6 above)
Comment (mactrz) [reply src/test/java/com/kainos/ea/backend/services/BandCompetencyServiceTest.java:27]: Changed

**Review by sebastianczajkowski — APPROVED** [review]

**Pull request merged**